java/com/google/re2j/InstBuilder.java

```diff
@@ -0,0 +1,33 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Original Go source here:
+// http://code.google.com/p/go/source/browse/src/pkg/regexp/syntax/prog.go
+
+package com.google.re2j;
+
+/**
+ * An instruction builder
+ * @see http://swtch.com/~rsc/regexp/regexp2.html
+ */
+final class InstBuilder {
+  
+
+  int op;
+  int out;  // all but MATCH, FAIL
+  int arg;  // ALT, ALT_MATCH, CAPTURE, EMPTY_WIDTH
+  int[] runes;  // length==1 => exact match
+                // otherwise a list of [lo,hi] pairs.  hi is *inclusive*.
+                // REVIEWERS: why not half-open intervals?
+
+  InstBuilder(int op) {
+    this.op = op;
+  }
+
+
+
+  Inst build() {
+    return new Inst(op, out, arg, runes);
```

Comment:
What's the value in having two representations for instructions, one before mergedOp is computed and one after?  This approach causes Prog.build to allocate more memory for no obvious gain.

Reply:
I wanted to make sure that Inst does not get modified after construction, 
so that I can still directly exposed the field as final. 
from an alloc perspective, the ones who really care about it - like me - will cache the compiled version. I originally did it in the same object but... then it's too easy for somebody else to add an optimisation somewhere to change the op on the Int and forget to recalculate the merge value.

Reply:
I'm not sure that's a strong enough justification for InstBuilder.  Insts are created in only one place.  Also, there are only three calls to the (merged) op() function today, and two of them could be simplified to a boolean predicate `op() == Inst.Op.RUNE`.  The third appears in a switch (in Thread.add), which seems confused and suboptimal since it does a switch to merge the rune cases then another switch to dispatch on the cases.

I suggest you get rid of op() and replace it by a function `boolean isRuneOp(op int)`, and get rid of InstBuilder too.